clis/teliod/src/main.rs

```diff
@@ -0,0 +1,240 @@
+//! Main and implementation of config and commands for Teliod - simple telio daemon for Linux and OpenWRT
+
+use anyhow::{bail, Result};
+use clap::Parser;
+use nix::libc::SIGTERM;
+use signal_hook_tokio::Signals;
+use std::{
+    fs::{self, File},
+    io::Write,
+    path::{Path, PathBuf},
+    process::id,
+    str::FromStr,
+};
+use tracing::{error, info, level_filters::LevelFilter};
+
+mod comms;
+mod daemon;
+
+use crate::{comms::DaemonSocket, daemon::Daemon};
+
+use telio::{device::Device, telio_model::features::Features, telio_utils::select};
+
+use serde::{de, Deserialize, Deserializer, Serialize};
+
+use futures::stream::StreamExt;
+
+struct TeliodDaemon;
+
+impl Daemon for TeliodDaemon {
+    const NAME: &'static str = "teliod";
+
+    fn get_wd_path() -> Result<PathBuf> {
+        if Path::new("/run").exists() {
+            Ok(PathBuf::from("/run/"))
+        } else {
+            Ok(PathBuf::from("/var/run/"))
+        }
+    }
+}
+
+impl TeliodDaemon {
+    /// Inits the daemon on the same process
+    fn init() -> Result<()> {
+        // Create pid file
+        let pid = id();
+        let mut pid_file = File::create(Self::get_pid_path()?)?;
+        pid_file.write_fmt(format_args!("{}", pid))?;
+        Ok(())
+    }
+}
+
+#[derive(Deserialize)]
+struct TeliodDaemonConfig {
+    #[serde(deserialize_with = "deserialize_log_level")]
+    log_level: LevelFilter,
+    log_file_path: String,
+}
+
+fn deserialize_log_level<'de, D>(deserializer: D) -> Result<LevelFilter, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    Deserialize::deserialize(deserializer).and_then(|s: String| {
+        LevelFilter::from_str(&s).map_err(|_| {
+            de::Error::unknown_variant(&s, &["error", "warn", "info", "debug", "trace", "off"])
+        })
+    })
+}
+
+#[derive(Parser)]
+#[clap()]
+#[derive(Serialize, Deserialize)]
+enum Cmd {
+    #[clap(about = "Runs the teliod event loop")]
+    Daemon { config_path: String },
+    #[clap(
+        about = "Forces daemon to add a log, added for testing to be, to be removed when the daemon will be more mature"
+    )]
+    HelloWorld { name: String },
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    match Cmd::parse() {
+        Cmd::Daemon { config_path } => {
+            if TeliodDaemon::is_running()? {
+                eprintln!("Teliod is already running");
```

Comment:
Note that in case daemon exits with error condition -> we should make sure that its exit code is non-zero. I guess we can use `anyhow::bail` or `return Err(...)` which nicely "represents" error condition.